buid(aio): various doc-gen fixes

petebacondarwin opened the pull request:
Fixes #15292
Fixes #15293
Fixes #14851
Fixes #15270

petebacondarwin committed:
build(aio): do not render private classes and members

petebacondarwin committed:
build(aio): do not render ignored docs

petebacondarwin committed:
build(aio): use forked version of Rho rendering engine

The original Rho is too strict when it comes to markdown headings.
It requires that there be a blank line separating the heading and the
next paragraph.  The forked version here fixes that; but the Rho project
will not merge it as it goes against there basic rules.

mary-poppins commented:
The angular.io preview for fd38caa is available [here][1].

[1]: https://pr15285-fd38caaecae7c24f4f1495edd48dff53d279c884.ngbuilds.io/

petebacondarwin committed:
build(aio): ensure that internal document links work with base href

petebacondarwin committed:
build(aio): remove cheatsheet processing and content

This will be replaced by a single file migrated from angular.io

petebacondarwin committed:
build(aio): don't process unnecessary example files in doc-gen

This lowers the `yarn docs` processing time by about 40% (from 50 secs
to 30 secs).

mary-poppins commented:
The angular.io preview for d06de26 is available [here][1].

[1]: https://pr15285-d06de2631313431ecc1e26ebcfe1284d2c697bf1.ngbuilds.io/

gkalpak commented on aio/transforms/angular.io-package/processors/processNavigationMap.js:
```diff
@@ -0,0 +1,48 @@
+  for(const key in node) {
+    const child = node[key];
+    if (key === 'url') {
+      const url = child.replace(/#\w*$/, ''); // strip hash
```
Why `/w`? This won't match `-` for example.

petebacondarwin replied:
I am rubbish at regex. Can you suggest a better one?

petebacondarwin replied:
OK, I am going with the full-on: `/#.*$/`

gkalpak replied:
Since hash is the last part of a URL, `/#.*/` is the first thing that comes to mind.

gkalpak commented on aio/transforms/angular.io-package/processors/processNavigationMap.js:
```diff
@@ -0,0 +1,48 @@
+    if (key === 'url') {
+      const url = child.replace(/#\w*$/, ''); // strip hash
+      if (isRelative(url) && !map[url]) {
+        errors.push({ path: path.join('.'), url});
```
<sub>😱 Inconsistent spacing 😱 around `{...}` 😱</sub>

petebacondarwin replied:
Gadzooks!

gkalpak commented on aio/transforms/angular.io-package/processors/processNavigationMap.js:
```diff
@@ -0,0 +1,48 @@
+      const navigationDoc = docs.find(doc => doc.docType === 'navigation-map');
+
+      if (!navigationDoc) {
+        log.warn('Missing navigation map document (docType="navigation-map"). Did you forget to add it to the readFileProcessor?');
```
Should this be an error instead of a warning?

petebacondarwin replied:
eventually. but right now we are not stopping the build for this.

gkalpak commented on aio/transforms/angular.io-package/processors/processNavigationMap.js:
```diff
@@ -0,0 +1,48 @@
+
+      if (!navigationDoc) {
+        log.warn('Missing navigation map document (docType="navigation-map"). Did you forget to add it to the readFileProcessor?');
+      }
```
Is it intentional that processing continues even if `navigationDoc` was not found?

petebacondarwin replied:
I guess I am being too lenient here :-)

gkalpak replied:
It will throw errors down the road, when trying to find properties on undefined. The question is, do you want it to throw those errors, or do you want to stop/handle the processing here :smiley:

gkalpak commented on aio/src/app/navigation/navigation.service.ts:
```diff
@@ -75,7 +75,7 @@ export class NavigationService {
   private computeUrlToNodesMap(navigation: NavigationViews) {
     const navMap = {};
-    Object.keys(navigation).forEach(key => navigation[key].forEach(node => walkNodes(node)));
+    Object.keys(navigation).forEach(key => Array.isArray(navigation[key]) && navigation[key].forEach(node => walkNodes(node)));
```
This seems a little confusing. A comment about what a non-Array value might be (or why it is OK to expect extra, non-Array values) would help.

petebacondarwin replied:
Yeah, you are right.
Actually what I ought to be doing here is extracting the version data and making it available.

gkalpak reviewed (changes requested):
Generally LGTM.
<sub>(AFAICT, PR only partly addresses #15270. The other part would be moving the cheatsheet over, right?)</sub>

petebacondarwin commented on aio/transforms/angular.io-package/processors/filterPrivateDocs.js:
```diff
@@ -0,0 +1,9 @@
+    $runAfter: ['extra-docs-added'],
+    $runBefore: ['computing-paths'],
+    $process: function(docs) {
+      return docs.filter(function(doc) { return !doc.name || (doc.name.indexOf('ɵ') !== 0); });
```
Well, the thing is that there can also be "member" docs that are members of a class.
I was trying to capture those as well...

gkalpak replied:
So, this processor and `marlBarredODocsAsPrivate` receive a different `docs` array.
(You can tell I don't know much about the docs-gen pipeline :grin:)

petebacondarwin replied:
I see that you know more than me :-). I'll change this

petebacondarwin committed:
build(aio): add version into navigation.json

The navigation.json is now passed through the dgeni pipeline.
The source file has been moved to `aio/content/navigation.json`
but the generated file will now appear where the original source file
was found, `aio/src/content/navigation.json`.

Everything inside `aio/src/content` is now generated and ignored by git.

The `processNavigationMap` processor in this commit adds the current version
information to the navigation.json file and verifies the relative urls in
the file map to real documents.

The navigationService exposes the versionInfo as an observable, which the
AppComponent renders at the top of the sidenav.

petebacondarwin committed:
test(aio): use MockLogger in AppComponent tests

This component sends a lot of messages to the console, which makes the
test run less easy to follow.

petebacondarwin commented:
@gkalpak I have fixed all of your comments.
Regarding the cheatsheet stuff. The migrator has been modified to bring over the new standalone content page. I was waiting for this PR to land before hitting everyone with an update to the content folder, which would contain numerous file changes.

mary-poppins commented:
The angular.io preview for aa099ee is available [here][1].

[1]: https://pr15285-aa099ee6e3a9e77f7e7e78dc3c696e11b26d85fe.ngbuilds.io/

mary-poppins commented:
The angular.io preview for f559cf2 is available [here][1].

[1]: https://pr15285-f559cf26a541488eba7d05c7b0129eb95221e2af.ngbuilds.io/

gkalpak commented on aio/src/app/navigation/navigation.service.ts:
```diff
@@ -43,10 +77,22 @@ export class NavigationService {
+  }
+
+  private getVersionInfo(navigationInfo: Observable<NavigationResponse>) {
+    const versionInfo = navigationInfo.map(response => response['__versionInfo']).publishReplay(1);
```
Is this replublishing necessary (given we already have the `navigationInfo` observable)?
Wouldn't the following work just as well:

```ts
private getVersionInfo(navigationInfo: Observable<NavigationResponse>) {
  return navigationInfo.map(response => respons['__versionInfo']);
}
```

gkalpak replied:
Is this replublishing necessary (given we already have the `navigationInfo` observable)?
Wouldn't the following work just as well:

```ts
private getVersionInfo(navigationInfo: Observable<NavigationResponse>) {
  return navigationInfo.map(response => respons['__versionInfo']);
}
```

(Same for `getNavigationViews` below.)

petebacondarwin replied:
If you don't republish then every subscriber will trigger the `map` function again.
Either way, I don't believe this is a significant performance issue for this one.
For the other one, I would rather not be calling the `unpluck` for every subscriber, so I made both published to be consistent.

petebacondarwin committed:
test(aio): fix test descriptions for ordering of processors

petebacondarwin commented on aio/src/app/navigation/navigation.service.spec.ts:
```diff
@@ -125,4 +125,23 @@ describe('NavigationService', () => {
+
+      const backend = injector.get(ConnectionBackend);
+      backend.connectionsArray[0].mockRespond(createResponse({
+        ['__versionInfo']: { raw: '4.0.0' }
```
'cos :-P

petebacondarwin commented on aio/src/app/navigation/navigation.service.spec.ts:
```diff
@@ -125,4 +125,23 @@ describe('NavigationService', () => {
+    });
+
+    it('should extract the version info', () => {
+      const backend = injector.get(ConnectionBackend);
```
nope

petebacondarwin commented on aio/src/app/navigation/navigation.service.ts:
```diff
@@ -12,28 +12,62 @@ import { LocationService } from 'app/shared/location.service';
 import { NavigationNode } from './navigation-node';
 export { NavigationNode } from './navigation-node';
 
+export type NavigationResponse = {'__versionInfo': VersionInfo } & { [name: string]: NavigationNode[] };
```
You're right. It turns out it worked, since we never actually create an explicit instance of this type. But since it is a bit wrong, I have changed it to 

```
export type NavigationResponse = {__versionInfo: VersionInfo } & { [name: string]: NavigationNode[]|VersionInfo };
```

which is not as accurate at describing the type but will work for assignment. You just need to do a little bit more typecasting later, is all.

petebacondarwin committed:
build(aio): ensure that internal document links work with base href

petebacondarwin committed:
build(aio): add version into navigation.json

The navigation.json is now passed through the dgeni pipeline.
The source file has been moved to `aio/content/navigation.json`
but the generated file will now appear where the original source file
was found, `aio/src/content/navigation.json`.

Everything inside `aio/src/content` is now generated and ignored by git.

The `processNavigationMap` processor in this commit adds the current version
information to the navigation.json file and verifies the relative urls in
the file map to real documents.

The navigationService exposes the versionInfo as an observable, which the
AppComponent renders at the top of the sidenav.

mary-poppins commented:
The angular.io preview for dec731e is available [here][1].

[1]: https://pr15285-dec731e3921bb1936e910a3816a0ba9dc77137b5.ngbuilds.io/

gkalpak reviewed (approved):
LGTM as soon as it LGT Travis.

mary-poppins commented:
The angular.io preview for a5038be is available [here][1].

[1]: https://pr15285-a5038be0df500ffd104226fcc5568659f1a99fba.ngbuilds.io/
